### PR TITLE
Added definition and test for when.unfold

### PR DESCRIPTION
**Pull request by mnahkies**

### CODE REVIEW

**Comment by vvakame** [comment]
*when/when.d.ts*

to authors(@derekcicerone @Nemo157). could you review this PR?
:+1: or :-1:?

check list

* [ ] pass the Travic-CI test?

File: when/when.d.ts [hunk 1]
```diff
@@ -142,6 +142,23 @@ declare module When {
...
+     * @param unspool function that, given a seed, returns a [valueToSendToHandler, newSeed] pair.
+     * May return an array, array of promises, promise for an array, or promise for an array of promises.
+     * @param predicate function that receives the current seed, and should return truthy when the unfold should stop
+     * @param handler function that receives the valueToSendToHandler of the current iteration.
+     * This function can process valueToSendToHandler in whatever way you need.
+     * It may return a promise to delay the next iteration of the unfold.
+     * @param seed initial value provided to the first unspool invocation. May be a promise.
+     */
+    function unfold<T, U>(unspool: (seed: U) => [T | Promise<T>, U | Promise<U>] | Promise<[T | Promise<T>, U | Promise<U>]>,
+                          predicate: (value: U) => boolean | Promise<boolean>,
```
Comment (Nemo157) [inline when/when.d.ts]: From https://github.com/cujojs/when/blob/master/docs/api.md#whenunfold
> `predicate` - function that receives the current seed, and should return truthy when the unfold should stop

Not sure if this is allowed to be `Promise<boolean>` or not, the docs don't explicitly allow it.

**Comment by Nemo157** [comment]
Other than the line note I added :+1:

**Comment by derekcicerone** [comment]
Sorry, I haven't used when in ages :(

**Comment by mnahkies** [comment]
@Nemo157 Good spotting. The code resolves the predicate though, and there is a test explicitly testing that this is the case:

https://github.com/cujojs/when/blob/ccd6970e55006abe73c110fe50eb315f3099bdf9/lib/decorators/iterate.js#L51

https://github.com/cujojs/when/blob/ccd6970e55006abe73c110fe50eb315f3099bdf9/test/iterate-test.js#L96

Interestingly the code has unfold marked as deprecated, but the api documentation doesn't mention this at all. I've left a note on the original issue regarding this: https://github.com/cujojs/when/issues/370#issuecomment-151386381

**Comment by Nemo157** [comment]
:+1:

**Comment by vvakame** [comment]
thanks mate!

**Pull request merged**